### PR TITLE
WIP: Replicate stack gan

### PR DESCRIPTION
**Pull request by Devin-Taylor**
DO NOT MERGE

### CODE REVIEW

File: models/generators.py [hunk 1]
```diff
@@ -33,15 +35,18 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
...
         self.upsample1 = UpSampling2D()
         self.conv1 = Conv2D(filters=num_filters, kernel_size=kernel_size, activation='relu', padding='same')
+        self.bn1 = BatchNormalization()
         self.upsample2 = UpSampling2D()
         self.conv2 = Conv2D(filters=num_output_channels, kernel_size=kernel_size, activation='tanh', padding='same')
 
     @tf.function
-    def call(self, noise):
-        x = self.dense1(noise)
+    def call(self, noise, embedding):
```
Comment (Devin-Taylor) [inline models/generators.py]: @alecokas can I assume embedding will be passed in here?

Comment (alecokas) [reply models/generators.py]: Yeah I think that makes sense

File: models/conditional_gans.py [hunk 2]
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
...
+            kernel_size=kernel_size,
+            num_filters=num_filters,
+            reshape_dims=reshape_dims
+        )
+
+        discriminator = DiscriminatorStage1(
+            img_size=img_size,
+            kernel_size=kernel_size,
+            num_filters=num_filters
+        )
```
Comment (alecokas) [inline models/conditional_gans.py]: Why are these before calling `super`?

Comment (Devin-Taylor) [reply models/conditional_gans.py]: they need to be passed into super because each stage has different generators and discriminators. Tbh I do not know if we really need the parent class here, but I just left it for now.

File: utils/data_helpers.py [hunk 3]
```diff
@@ -137,11 +140,21 @@ def download_flowers():
...
+        old_size = image.size[:2]
+        ratio = max(image_dims)/max(old_size)
+        new_size = tuple([int(x*ratio) for x in old_size])
+        image = image.resize(new_size, Image.ANTIALIAS)
+        new_img = Image.new('RGB', image_dims)
+        new_img.paste(image, ((image_dims[0]-new_size[0])//2,
+                              (image_dims[1]-new_size[1])//2))
+        img_buffer = io.BytesIO()
+        new_img.save(img_buffer, format='PNG')
+        image_string = img_buffer.getvalue()
```
Comment (alecokas) [inline utils/data_helpers.py]: I assume this is the code that maintains the aspect ratio?

File: utils/utils.py [hunk 4]
```diff
@@ -53,7 +53,7 @@ def sample_normal(mean, log_var):
             latent_dim)
     """
     std = tf.math.exp(0.5 * log_var)
-    epsilon = tf.random.normal(std)
+    epsilon = tf.random.normal(tf.shape(std)) # TODO is shape the correct thing to use here?
```
Comment (alecokas) [inline utils/utils.py]: What do you mean?

File: utils/data_helpers.py (hunk 3 above)
Comment (Devin-Taylor) [reply utils/data_helpers.py]: yup, pretty much:
1) determine aspect ratio
2) determine reduced image size required to maintain ar
3) create blank square image
4) paste resized image in middle of blank image

File: utils/utils.py [hunk 5]
```diff
@@ -82,7 +82,8 @@ def save_options(options, save_dir):
     for option in vars(options):
         opt_dict[option] = getattr(options, option)
 
-    mkdir(save_dir)
+    if not os.path.exists(save_dir):
+        mkdir(save_dir)
```
Comment (alecokas) [inline utils/utils.py]: `mkdir` has a check built into it, so you don't need to check before calling the function

File: models/conditional_gans.py (hunk 2 above)
Comment (alecokas) [reply models/conditional_gans.py]: OK, sounds good :+1:

File: utils/utils.py (hunk 4 above)
Comment (Devin-Taylor) [reply utils/utils.py]: I think it is ok, but I wanted to double check with the paper when I had a chance

File: utils/utils.py (hunk 5 above)
Comment (Devin-Taylor) [reply utils/utils.py]: ah true, will edit it

File: utils/data_helpers.py (hunk 3 above)
Comment (alecokas) [reply utils/data_helpers.py]: :+1:

File: utils/utils.py (hunk 4 above)
Comment (alecokas) [reply utils/utils.py]: Checked: this is correct

**Comment by Devin-Taylor** [comment]
@alecokas please review and merge when happy. Stage 1 implementation complete and running end-to-end. Please *do not* delete the branch.

Changes:
- resize images, maintaining aspect ratio, pad border to get square
- implemented stage 1 conditional augmentation
- implemented stage 1 generator
- implementation stage 1 discriminator
- implement generator and discriminator loss functions
- configured training loop
- integrated batched training

Notes:
- Not training perfectly (discriminator winning too easily) but will resolve that in a different PR.

File: models/generators.py [hunk 6]
```diff
@@ -50,4 +54,115 @@ def loss(self, predictions_on_fake):
...
+
+    def loss(self, predictions_on_fake, mean, log_sigma):
+        """ Only calculate the loss based on the discriminator
+            predictions for the images generated by this model.
+        """
+        kl_coeff = 2
+        return self.xent_loss_fn(predictions_on_fake, tf.ones_like(predictions_on_fake)) + kl_coeff * self.kl_loss(mean, log_sigma)
+
+    def kl_loss(self, mean, log_sigma):
+        loss = -log_sigma + .5 * (-1 + tf.exp(2. * log_sigma) + tf.square(mean))
```
Comment (alecokas) [inline models/generators.py:138]: Assuming this is correct for now lol

File: trainers/text_to_image_trainer.py [hunk 7]
```diff
@@ -1,7 +1,10 @@
 import io
+
 import numpy as np
+import tensorflow as tf
 from PIL import Image
 from tqdm import trange
+
 from trainers.base_trainer import Trainer
```
Comment (alecokas) [inline trainers/text_to_image_trainer.py:9]: Why the spaces?

File: models/discriminators.py [hunk 8]
```diff
@@ -48,4 +52,111 @@ def loss(self, predictions_on_real, predictions_on_fake):
...
+        super().__init__()
+        self.img_size = img_size
+        num_channels = self.img_size[0]
+        self.xent_loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=False)
+        self.optimiser = tf.keras.optimizers.Adam(1e-4)
+        # TODO: Add the correct layers as per the paper
+        self.conv1 = Conv2D(filters=64, kernel_size=(4, 4), strides=(2, 2))
+        self.leaky_relu_1 = LeakyReLU()
+
+        self.conv2 = Conv2D(filters=64*2, kernel_size=(4, 4), strides=(2, 2))
```
[the remark this answers is not in the document]
Comment (Devin-Taylor) [reply models/discriminators.py:77]: haha na it shouts at me for that, you using vscode default python extension?

File: models/conditional_gans.py [hunk 9]
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
...
+                text embedding. Shape (batch_size, feature_size, embedding_size)
+
+            Returns
+            Tensor
+                learnt mean of embedding distribution. Shape (batch_size, reshape_dims/2)
+            Tensor
+                learnt log variance of embedding distribution. Shape (batch_size, reshape_dims/2)
+
+        """
+        x = self.flatten(embedding)
```
Comment (alecokas) [inline models/conditional_gans.py:113]: So we take the mean and var over all elements? Across all samples, all features, and all embedding dimensions?

File: trainers/text_to_image_trainer.py (hunk 7 above)
Comment (Devin-Taylor) [reply trainers/text_to_image_trainer.py:9]: right click + sort imports, do not as me ¯\_(ツ)_/¯

File: models/conditional_gans.py [hunk 10]
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
...
+            embedding : Tensor
+                text embedding. Shape (batch_size, feature_size, embedding_size)
+
+            Returns
+            Tensor
+                sampled embedding. Shape (batch_size, reshape_dims/2)
+        """
+        mean, sigma = self.generate_conditionals(embedding)
+        smoothed_embedding = sample_normal(mean, sigma)
+        return smoothed_embedding, mean, sigma
```
Comment (alecokas) [inline models/conditional_gans.py:130]: So do I understand correctly that conditional augmentation is simply sampling a normal with the mean and var of all the elements across all dims in the batch? I'm asking because I can't really see a random variable that is being conditioned on. I will have to re-read the papers, but I thought worth checking because it doesn't seem intuitive to me right now

File: trainers/text_to_image_trainer.py [hunk 11]
```diff
@@ -30,16 +33,54 @@ def train_epoch(self, train_loader, epoch_num):
...
                 # For Caption: text_tensor = np.frombuffer(sample['text'].numpy(), dtype=np.float32).reshape(10, 1024)
-                # print(text_tensor)
-                # name = sample['name'].numpy().decode("utf-8")
                 # label = sample['label'].numpy()
+
+                noise_z = tf.random.normal([batch_size, 100])
+
+                with tf.GradientTape() as generator_tape, tf.GradientTape() as discriminator_tape:
+                    smoothed_embedding, mean, log_sigma = self.model.conditional_augmentation(text_tensor)
+                    embedding_z = tf.concat([smoothed_embedding, noise_z], 1)
```
Comment (alecokas) [inline trainers/text_to_image_trainer.py:55]: Should we be doing this outside the gradient tape maybe?

Comment (alecokas) [reply trainers/text_to_image_trainer.py:55]: Oh wait never mind, we have some linear layers in there that need to be backpropped.

File: models/conditional_gans.py [hunk 12]
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
...
+            num_latent_dims=num_latent_dims,
+            kernel_size=kernel_size,
+            num_filters=num_filters,
+            reshape_dims=reshape_dims
+        )
+
+    @tf.function
+    def call(self, embedding, z):
+        """
+        """
```
Comment (alecokas) [inline models/conditional_gans.py:165]: Maybe just add `None` or raise `NotImplementedError` - I think this is a syntax error atm.

File: models/discriminators.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alecokas) [reply models/discriminators.py:77]: I usually use flake8 - I'm fine with this as is. Just curious because flake8 complains I think

File: trainers/text_to_image_trainer.py (hunk 7 above)
Comment (alecokas) [reply trainers/text_to_image_trainer.py:9]: :joy:

File: models/conditional_gans.py (hunk 9 above)
Comment (Devin-Taylor) [reply models/conditional_gans.py:113]: yeah, technically. Might be worth averaging over the sample dimension first then flattening?

Comment (alecokas) [reply models/conditional_gans.py:113]: Yeah we should double check this

File: models/conditional_gans.py (hunk 10 above)
Comment (Devin-Taylor) [reply models/conditional_gans.py:130]: Screenshot from 2020-04-06 23-21-53

File: trainers/text_to_image_trainer.py (hunk 11 above)
Comment (Devin-Taylor) [reply trainers/text_to_image_trainer.py:55]: conditional augmentation has the linear layers, so are trainable. Which means they need to be on the tape, no?

File: models/conditional_gans.py (hunk 12 above)
Comment (Devin-Taylor) [reply models/conditional_gans.py:165]: cool, fixed

File: models/conditional_gans.py (hunk 9 above)
Comment (Devin-Taylor) [reply models/conditional_gans.py:113]: hard to tell, discussions on the multiple samples are vague. I think there was an older paper that mentions it, will dig it up and see what they do
